ipaserver/rpcserver.py

```diff
@@ -872,33 +871,15 @@ def __call__(self, environ, start_response):
             return self.bad_request(environ, start_response, "no user specified")
 
         # allows login in the form user@SERVER_REALM or user@server_realm
-        # FIXME: uppercasing may be removed when better handling of UPN
-        #        is introduced
-
-        parts = normalize_name(user)
-
-        if "domain" in parts:
-            # username is of the form user@SERVER_REALM or user@server_realm
-
-            # check whether the realm is server's realm
-            # Users from other realms are not supported
-            # (they do not have necessary LDAP entry, LDAP connect will fail)
-
-            if parts["domain"].upper()==self.api.env.realm:
-                user=parts["name"]
-            else:
-                return self.unauthorized(environ, start_response, '', 'denied')
-
-        elif "flatname" in parts:
-            # username is of the form NetBIOS\user
-            return self.unauthorized(environ, start_response, '', 'denied')
-
-        else:
+        try:
+            user_principal = kerberos.Principal(user)
+        except Exception:
             # username is of the form user or of some wild form, e.g.
-            # user@REALM1@REALM2 or NetBIOS1\NetBIOS2\user (see normalize_name)
+            # user@REALM1@REALM2 or NetBIOS1\NetBIOS2\user
+            return self.unauthorized(environ, start_response, '', 'denied')
 
-            # wild form username will fail at kinit, so nothing needs to be done
-            pass
+        if not (user_principal.is_user or user_principal.is_enterprise):
```

Comment:
This code actually prevents services to contact us. you should also allow user_principal.is_service.